# Project for Programming for Data Analysis

## Project Summary

In this project I have choosen a real world scenario do to some investigation using the numpy.random function in order to simulate a dataset, trying to match as closely as possible the real world scenario. I will be detailing my investigation and show the simulated data.

### How to be accepted in a University in the USA

For this project I have choosen the scenario of the how students are accepted by universities in the United States. After research, I have come to conclude that the most important variable related to the acceptance of a student in a University is the sudent's SAT score, followed by academic performance, financial statements and student's references, but before going any further I will go over in more depth on what exactly each variable is and how they are calculated, starting with the SAT scores.

#### SAT:

SAT means Scholastic Aptitude Test or Scholastic Assessment Test and was developed and published by the College Board, a private, non-profit organization in the United States. and It's administered by the Educational Testing Service.

The test is intended to assess students' readiness for college, the current SAT, introduced in 2016, takes three hours to finish, plus 50 minutes for the SAT with essay.

The SAT is composed by two sections – Evidence Based Reading and Writing (EBRW), and Math. Students can earn a scaled score of between 200 and 800 points on each section, for a total of 1600 points.

Now, in order to determine what makes for good SAT scores, it’s important to understand how SAT scoring works. A student's total scoure out of 1600 corresponds to a percentile ranking. The percentile tells what percent of students scored better than others. So as an example, if a given student received a 70th-percentile score, he would have scored better than 70% of all test-takers.

After further research I have discovered that the SAT scores follow a normal distribution, meaning that students performance tends to cluster around the middle of the scale. Only fewer students score towards the higher and lower end of the scale.

Now we already know how the score is calculated, but in order to simulate a dataset as closely as possible with the reality we also need to know what are the average score and the minimum score that would be accepted by certain Universities.

After research under some Universities, a 1280 score would be an accepted score for schools like the University of Cincinnati (average SAT 1240), ASU (average SAT 1230), and Temple (average SAT 1220). But it would be a very low score for highly selective institutions like MIT, CalTech, Duke, University of Chicago, or Johns Hopkins, for example.

Although, some other schools are not super selective, and have lower average scores, a score like 1040 would still be a good score for schools like UT Rio Grande Valley (average SAT score 1020), Indiana University Northwest (average SAT score 1000) or Flagler College (average SAT score 1030).

The average SAT score can be calculated by states, but for the purpose of this simulation I will use the average of the  National SAT score, which is 1060, along with a standard deviation of 195. 

As for the purporse of the final calculation, this ficticial university would require a SAT score of 1060 to be accepted.


#### Academic performance:

As the SAT scores, the academic performance, or Grade point average (GPA) also has an impact on student's being accepted by a University, and after research on how Universities calculate the academic performance of a student, I have summarized here the proccess which I will be using for the data simulation:

Grades in the US are given in five levels of A, B, C, D and F, each with a grade point from 0 to 4. The academic performance is generally calculated by converting the grades for all the courses graded in the A-F range to grade points, and calculating the weighted average based on the number of credits. 

After some research, I've found that nationally, in the US, the average high school GPA is 3.0, which is considered a B grade. Typically a 3.5-4.0 GPA,  means an A- or A grade, which is expected for admission to top colleges, as an example, student's from Harvard have an average GPA of 3.93. 

For this particular simulation, I will consider the GPA as an normal distribution with an average GPA of 3.0 and standard deviation of 0.25, as for the final passing grade to be accepted, I will be considering it a GPA score of at least 3.0.


#### Financial Statements:

Another factor that is important for a student being accepted is to prove that he or she will be able to pay the tuition fee and living expenses for the entire duration of the studies. Universities may need student's to submit their sponsor’s or bank account statements for a particular period.

Therefore for this simulation I will use the average prices provided from College Board for Public Universities in a four-year course with in-state fees, which are \$9,970 for the tuition fee, plus the value of the room and boarding of an additional \$10,800 which leaves to a total of \$20,770

This variable will be called "fin" and in order for students being accepted, they would need to provide statements with at least \$20,770, otherwise they would have their application denied.

#### References:

This latst factor is also very important, and is required of every applicant that he or she sends reference letters of recommendation from tutors at their school. This will provide the admissions committees with valuable insight into the student's abilities and character, and besides the references each student is supposed to send along an essay, which together provides an important record of the student's achievements.

Reference letters serve one purpose: to convince those reading them that the student belong in their school.

Since the references are basically letters and are difficult to calculate a student's passing rate, for this particular simulation, this variable will be categorical with four possible values: "fail", "reasonable", "good" and "perfect", and all of the last three values would mean student's were accepted and the "fail" would be the only variable meaning that the student wasn't accepted.


### Data Simulation

In this section I will go over the simulation of this scenario, considering that there is 500 students applying for the same course in a given University, and in the end of analysis, I will calculate how many would be be accepted.

In [1]:
#importing numpy
import numpy as np

# Create data for sat scores using a normal distribution
np.random.seed(10) #used a seed here in order to always keep the same result
sat = np.random.normal(1060, 195, 500)
sat

#creating the dataframe for SAT scores
import pandas as pd 
sat_df = pd.DataFrame(sat, columns=['Score'])

#showing the first 5 scores
sat_df.head()

Score
0  1319.659368
1  1199.479400
2   758.646943
3  1058.365149
4  1181.160515

In [2]:
# Create data for academic performance
gpa = np.random.normal(3.0, 0.25, 500)
gpa

#creating the dataframe for GPA
gpa_df = pd.DataFrame(gpa, columns=['GPA'])

#showing the first 5 GPA's
gpa_df.head()

GPA
0  3.359813
1  3.345537
2  3.230330
3  2.856778
4  3.318762

In [3]:
#create data for financial statements
fin = np.random.normal(20.770, 0.25, 500)

#creating the dataframe for financial statements
fin_df = pd.DataFrame(fin, columns=['Financials'])

#showing the first 5 financial statements
fin_df.head()

Financials
0   21.148313
1   20.734417
2   20.735145
3   21.063877
4   21.065947

In [4]:
#create data for references
outcomes = ['fail', 'reasonable', 'good', 'perfect']
#created a probability for each string, leaving a higher probability for 'reasonable' and 'good'
prob = [0.1, 0.4, 0.4, 0.1]
ref = np.random.choice(outcomes, 500, p=prob)

##creating the dataframe for the references
ref_df = pd.DataFrame(ref, columns=['References'])

#showing the first 5 references
ref_df.head()

References
0  reasonable
1        good
2        good
3        good
4  reasonable

In [5]:
# merging all dataframes in one
merged = pd.concat([sat_df, gpa_df, fin_df, ref_df], axis=1)


#creating the dataframe for the final merged list
merged_df = pd.DataFrame(merged)

#save list to a csv file
merged_df.to_csv("merged_df.csv")

#displaying the list
merged_df

Score       GPA  Financials  References
0    1319.659368  3.359813   21.148313  reasonable
1    1199.479400  3.345537   20.734417        good
2     758.646943  3.230330   20.735145        good
3    1058.365149  2.856778   21.063877        good
4    1181.160515  3.318762   21.065947  reasonable
5     919.583316  2.917792   20.940852        good
6    1111.774759  2.963415   20.466222  reasonable
7    1081.166963  3.264609   20.902890        good
8    1060.836829  2.972230   20.831357  reasonable
9    1025.952959  2.990280   20.832997  reasonable
10   1144.440107  2.919409   21.206115        fail
11   1294.592288  2.833559   20.530475     perfect
12    871.812194  2.823845   21.226167  reasonable
13   1260.513445  2.911559   20.350014     perfect
14   1104.582875  3.037982   20.491398  reasonable
15   1146.801835  2.923950   20.784788        good
16    838.362569  2.705286   20.691075        good
17   1086.351691  2.962717   20.887608        good
18   1349.484715  2.943726   20.964098  reasonable
19    849.438047  3.017539   20.503844        good
20    674.342985  3.113017   20.724348     perfect
21    720.042402  3.073517   20.822862  reasonable
22   1111.883682  2.793122   20.436819  reasonable
23   1525.068629  2.792882   20.718108  reasonable
24   1279.119794  2.876873   20.768043        fail
25   1386.161332  2.812095   20.861317        fail
26   1079.334097  2.602905   20.669993  reasonable
27   1332.609294  2.882817   21.053131  reasonable
28   1007.106642  2.642220   20.745346        fail
29   1179.574816  2.684126   20.717244  reasonable
..           ...       ...         ...         ...
470   984.832021  2.792897   20.646994  reasonable
471  1364.240712  2.843675   20.994150     perfect
472  1361.306543  3.066697   21.359539        good
473   925.120907  3.131504   21.030004  reasonable
474  1228.581102  2.922977   20.796993  reasonable
475   650.930585  3.096009   20.817946  reasonable
476  1093.207465  3.112673   20.571332        good
477   912.223218  2.922641   20.933369  reasonable
478  1248.723031  2.391448   20.753713  reasonable
479   889.467616  3.008720   20.353563        fail
480  1260.428876  3.363505   20.860064        good
481   903.436522  2.864936   20.521130  reasonable
482   976.480836  3.185038   20.875627        good
483   912.693912  2.822928   21.325588        good
484  1102.996895  2.835568   20.555083  reasonable
485   792.242734  3.527775   20.837893  reasonable
486  1186.117876  2.981758   20.877635        good
487   928.667146  2.764402   20.669079        good
488  1497.163968  3.286252   21.037800        fail
489  1011.935011  3.077719   20.996062        fail
490   911.097154  3.098033   20.723577        good
491  1153.753805  2.526352   20.662639     perfect
492   897.799155  2.914622   20.526792  reasonable
493  1175.671900  3.017443   21.127623     perfect
494   972.488068  2.959911   20.752352        good
495  1155.100708  2.746415   20.970713  reasonable
496  1070.981372  2.917043   20.694592        fail
497  1166.427971  3.360174   20.630525  reasonable
498   926.791132  2.902446   20.462465  reasonable
499   931.491625  3.160613   20.678772        good

[500 rows x 4 columns]

### Calculating how many students were actually accepted out of the 500

In [6]:
#filtering students that had a SAT score equal or higher than 1060 
accepted_df = merged_df[merged_df.Score >= 1060]
#filtering students that had a GPA score equal or higher than 3
accepted_df = accepted_df[accepted_df.GPA >= 3.0]
#filtering students that had financial statements equally or higher than 20.770
accepted_df = accepted_df[accepted_df.Financials >= 20.770]
#filtering students that had references 'good', 'reasonable' or 'perfect'
accepted_df = accepted_df[accepted_df.References.str.contains("good|reasonable|perfect")]

#displaying final accepted students
accepted_df

Score       GPA  Financials  References
0    1319.659368  3.359813   21.148313  reasonable
4    1181.160515  3.318762   21.065947  reasonable
7    1081.166963  3.264609   20.902890        good
38   1305.012091  3.131135   21.069907        good
47   1120.873138  3.242640   20.791473        good
51   1116.373370  3.036758   20.883675  reasonable
53   1198.091204  3.054640   21.069045  reasonable
66   1264.005657  3.411190   20.974572     perfect
71   1149.876066  3.125224   21.178626  reasonable
88   1102.366703  3.518762   20.858153  reasonable
91   1107.686075  3.027386   20.949197  reasonable
94   1198.832008  3.064403   21.113343        good
95   1239.062485  3.328231   21.013730        good
100  1082.907754  3.142673   20.948909  reasonable
108  1197.831939  3.085682   20.783694  reasonable
112  1528.843493  3.120754   21.114835  reasonable
118  1113.244617  3.232278   20.948709        good
123  1540.738391  3.119412   21.151973  reasonable
124  1329.730285  3.007350   20.998002        good
125  1127.515457  3.040453   20.853833  reasonable
128  1383.059742  3.134567   20.910040  reasonable
145  1374.809555  3.467497   20.996193        good
149  1281.319769  3.018915   21.066139  reasonable
163  1189.162932  3.275968   20.930807        good
196  1244.689585  3.551221   20.857611  reasonable
200  1085.961812  3.226623   20.946490     perfect
201  1294.535055  3.249814   20.950747  reasonable
222  1194.141422  3.188639   20.897969        good
233  1108.589838  3.053446   21.248525  reasonable
239  1275.587152  3.264704   20.812108        good
241  1165.405439  3.043785   20.872423     perfect
248  1386.444731  3.358475   20.821208        good
254  1073.416287  3.537993   21.021696        good
260  1430.949288  3.153495   20.855363        good
271  1060.784858  3.171712   20.844794  reasonable
276  1257.727003  3.325067   20.895601  reasonable
283  1111.556965  3.279508   20.799219  reasonable
292  1158.265746  3.274280   20.872982  reasonable
301  1231.499710  3.039093   20.783067        good
307  1106.671157  3.313720   20.868056  reasonable
320  1274.721560  3.119399   20.915724        good
323  1285.820056  3.209409   20.954598  reasonable
328  1270.547482  3.075393   20.859814        good
366  1098.468084  3.012186   20.803705  reasonable
384  1208.096910  3.117374   20.897897  reasonable
393  1116.123035  3.119286   21.159915        good
403  1144.912826  3.017935   20.889672        good
434  1160.492323  3.027358   20.889106  reasonable
435  1524.904301  3.053819   21.155383        good
436  1163.192102  3.430097   21.047083  reasonable
455  1345.802726  3.021246   20.906831  reasonable
472  1361.306543  3.066697   21.359539        good
480  1260.428876  3.363505   20.860064        good
493  1175.671900  3.017443   21.127623     perfect

In [7]:
#showing total number of accepted students
accepted_df.shape[0]

54

In [8]:
#Calculating the percentage of accepted students
accepted_df.shape[0] / merged_df.shape[0] * 100

10.8

##### Conclusion
As demonstrated by the simulation above, from the total of 500 applicants, 54 students got accepted, which translates to a 10.8% acceptance rate.